Implement @pure-unless-callable-is-impure

zonuexe opened the pull request:
Implements the `@pure-unless-callable-is-impure` PHPDoc tag (parser support: phpstan/phpdoc-parser#253, merged): a parameter-level annotation declaring that the function/method is pure **unless** the callable passed to that parameter is impure. Rebuilt on top of `2.2.x`.

## How it works

- The tag (and its `@phpstan-` prefixed alias) is resolved into per-parameter flags, threaded through the reflection layer (`ExtendedParameterReflection::isPureUnlessCallableIsImpureParameter()`), and populated for builtins via `functionMetadata` (`array_filter`, `array_map`, `array_reduce`, the `array_*u*` comparator family, `call_user_func(_array)`, `forward_static_call(_array)`, `preg_replace_callback(_array)`, …) and for `array_reduce` via its stub.
- `SimpleImpurePoint::createFromVariant()` evaluates the purity of arguments passed to flagged parameters:
  - provably **pure** callback (or omitted / `null`) → **no impure point** — the call is treated as pure;
  - provably **impure** callback → the impure point becomes **certain**;
  - unknown callable (opaque `callable` type) → unchanged **possibly-impure** behavior.
- Since the no-effect detection (`NoopExpressionNode`) is driven by "the statement has zero impure points", standalone `array_map($pureCb, $arr);` statements are now reported by `CallToFunctionStatementWithoutSideEffectsRule` with no rule changes.

```php
/** @phpstan-pure */
function doubled(array $arr): array
{
    // no longer false positive "Possibly impure"
    return array_map(static fn (int $x): int => $x * 2, $arr);
}

// now: "Call to function array_map() on a separate line has no effect."
array_map('is_string', $arr);
// still allowed as a statement (impure callback)
array_map(static function ($x) { echo $x; }, $arr);
```

- [x] Add function map
- [x] Add tests (fail-first: `PureFunctionRule`/`PureMethodRule` regressions, `CallToFunctionStatementWithoutSideEffectsRule` cases incl. impure/opaque/by-ref non-regressions)
- [x] Add rules (no new rule needed — the impure-point suppression feeds both the pure-context check and the existing no-effect rule)

refs https://github.com/phpstan/phpstan-src/pull/3106, https://github.com/phpstan/phpstan-src/pull/5912

Closes https://github.com/phpstan/phpstan/issues/11101
Closes https://github.com/phpstan/phpstan/issues/11100

zonuexe committed:
wip

zonuexe committed:
Add pureUnlessCallableIsImpureParameters

zonuexe committed:
Add isPureUnlessCallableIsImpureParameter method

zonuexe commented on src/PhpDoc/PhpDocNodeResolver.php:
```diff
@@ -413,6 +414,24 @@ public function resolveParamImmediatelyInvokedCallable(PhpDocNode $phpDocNode):
+	public function resolveParamPureUnlessCallableIsImpure(PhpDocNode $phpDocNode): array
+	{
+		$parameters = [];
+		// TODO: implement phpstan/phpdoc-parser
```
Will be replaced by https://github.com/phpstan/phpdoc-parser/pull/253

zonuexe commented on src/Reflection/Native/NativeParameterWithPhpDocsReflection.php:
```diff
@@ -22,6 +22,7 @@ public function __construct(
 		private ?Type $outType,
 		private TrinaryLogic $immediatelyInvokedCallable,
 		private ?Type $closureThisType,
+		private bool $pureUnlessCallableIsImpureParameter = false,
```
WIP. `= false` will be removed.

ondrejmirtes commented:
Just updated phpdoc-parser here so you can use that.

zonuexe committed:
Replace with $phpDocNode->getPureUnlessCallableIsImpureTagValues()

zonuexe committed:
Add pureUnlessCallableIsImpureParameters to functionMetadata

zonuexe committed:
Add pureUnlessCallableIsImpureParameters to functionMetadata_original.php

staabm commented:
@zonuexe any plans to move this forward?

zonuexe commented:
@staabm Hi, I know this branch is important but it was blocked due to my motivation. I'm going to rebase it on 2.0.x and do the necessary work. If I miss it tonight, I won't be able to start the rest of the work until next week, so it would be quicker for you to take over then.

staabm commented:
I don't want to put pressure on you. It was not clear to me whether you stopped working on it or not.

I am totally fine if this takes a bit more time - thank you

zonuexe commented:
No problem, this is a very interesting subject for me too, so I'm happy to work on it because I realized that there are other people who are interested in it. More importantly, I'm always helped by your hard work.

zonuexe committed:
Fix

VincentLanglet commented:
Friendly ping @zonuexe, still interested in this PR ? Do you have time to finish it ?
Do you look for help ? What is missing ?

zonuexe commented:
@VincentLanglet 
Thank you for the suggestion, sorry for making you wait so long.
I've been working on PHPStan 2.1.18 a few times over the last few weeks, and just finished a PHPStan workshop at a local PHP conference, so I'll get back to that tomorrow.

zonuexe committed:
Fix

zonuexe committed:
hasSideEffects default true

zonuexe committed:
wip

VincentLanglet commented:
Friendly ping @zonuexe 

I don't want to put pressure on you. It's just not clear to me whether you stopped working on it or not.
What is missing ?

It will unblock the https://github.com/phpstan/phpdoc-parser/pull/259 pr :)

VincentLanglet commented:
Maybe I'm wrong, but I start to think this annotation is useless
https://phpstan.org/r/d3af976b-17ef-4ca6-87cb-4a1325622cee

PHPStan already handle cb with immediately/later invoked callable, even for a pure function/method.

So we just need to declare array_map and so as pure.

devbanana commented:
i'm happy to help contribute towards a solution if it's decided which direction is best.

VincentLanglet commented:
Hi @devbanana feel free to work on this.

Laybe @zonuexe could explain what he had in mind ; or @ondrejmirtes will have some guidelines

zonuexe committed:
Thread pure-unless-callable-is-impure through the reflection layer

Add isPureUnlessCallableIsImpureParameter() on parameter reflections and getPureUnlessCallableIsImpureParameters() on function/method reflections, populated from PHPDoc and function metadata.

zonuexe committed:
Wire pure-unless-callable-is-impure into the analyser

Return the parameter flags from getPhpDocs(), thread them through enterClassMethod(), and consult them when collecting impure points for calls.

zonuexe committed:
Parse @pure-unless-callable-is-impure PHPDoc tag

Resolve the tag into per-parameter flags and merge them across parent PHPDocs.

zonuexe committed:
Mark callback-accepting builtins as pure-unless-callable-is-impure

array_filter/array_map/array_reduce, the array_u*/array_*_u* family, call_user_func(_array), forward_static_call(_array) and preg_replace_callback(_array) are pure unless their callback argument is impure.

zonuexe committed:
Add tests for @pure-unless-callable-is-impure

Type-inference fixture plus function-metadata expectations.

zonuexe committed:
Suppress call impure points when pure-unless-callable-is-impure callbacks are pure

SimpleImpurePoint::createFromVariant() now evaluates the purity of arguments passed to flagged parameters: provably pure (or omitted) callbacks produce no impure point, provably impure ones make the point certain, unknown callables keep the current possibly-impure behavior. This fixes the false positive for array_map() with a pure callback in @phpstan-pure functions and enables the no-effect report for standalone statements.

Closes https://github.com/phpstan/phpstan/issues/11100
Closes https://github.com/phpstan/phpstan/issues/11101

zonuexe committed:
Fix pure-unless-callable-is-impure parameter flag propagation

The flag was silently dropped wherever parameters are re-wrapped (combineAcceptors, overrideParameterType, resolved template variants, unresolved method prototypes) and the BetterReflection function path never read the PHPDoc tag. Also honor the @phpstan- prefixed alias, strip the leading $ from the tag's parameter name, read the tag from stub PHPDocs, and annotate array_reduce's stub.

zonuexe committed:
Fix CI: avoid named arguments, adjust no-effect fixture

Named arguments are not supported by the PHP 7.4 downgrade, so pass the enterClassMethod() parameters positionally. The file-without-errors.php fixture contained a standalone array_filter() call that is now correctly reported as having no effect - wrap it in echo count().

zonuexe committed:
Cover pure-unless-callable-is-impure verdict for maybe-null and maybe-callable arguments

Kills the two escaped Infection mutants in SimpleImpurePoint: a (callable|null) callback exercises the isNull() branch and a (pure-callable|int) callback exercises the isCallable() branch - both keep the call possibly impure.

staabm commented on resources/functionMetadata.php:
```diff
@@ -733,43 +733,47 @@
 	'array_diff_key' => ['hasSideEffects' => false],
-	'array_diff_uassoc' => ['hasSideEffects' => false],
-	'array_diff_ukey' => ['hasSideEffects' => false],
+	'array_diff_uassoc' => ['pureUnlessCallableIsImpureParameters' => ['key_compare_func' => true]],
```
since this file format got more complicated, I think it would make sense to describe at the file-level comment how the shape is.
we could even add a inline `@var` array shape definition right before the huge array.

(I did not verify whether we analyse this big file though)

staabm commented on src/Reflection/Php/PhpClassReflectionExtension.php:
```diff
@@ -642,7 +642,7 @@ private function createMethod(
 			$isPure = null;
 			if ($this->signatureMapProvider->hasMethodMetadata($declaringClassName, $methodReflection->getName())) {
-				$isPure = !$this->signatureMapProvider->getMethodMetadata($declaringClassName, $methodReflection->getName())['hasSideEffects'];
+				$isPure = !($this->signatureMapProvider->getMethodMetadata($declaringClassName, $methodReflection->getName())['hasSideEffects'] ?? true);
```
I feel this would be more readable when written in multiple lines

```suggestion
				$hasSideEffects = $this->signatureMapProvider->getMethodMetadata($declaringClassName, $methodReflection->getName())['hasSideEffects'] ?? true;
				$isPure = !$hasSideEffects;
```

otherwise its not 100% clear what this line defaults to

---

alternatively maybe even do it like a few lines below

```
$metadata = $this->signatureMapProvider->getMethodMetadata($declaringClassName, $methodReflection->getName());
$hasSideEffects = ['hasSideEffects'] ?? true;
$isPure = !$hasSideEffects;
 ```

staabm commented on src/Reflection/Php/PhpFunctionFromParserNodeReflection.php:
```diff
@@ -177,6 +179,12 @@ public function getParameters(): array
+				$pureUnlessCallableIsImpureParameter = $this->pureUnlessCallableIsImpureParameters[$parameter->var->name];
+			} else {
+				$pureUnlessCallableIsImpureParameter = false;
+			}
```
`??` instead?

staabm commented on src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php:
```diff
@@ -110,18 +110,30 @@ public function findFunctionReflection(string $functionName): ?NativeFunctionRef
 						$phpDocType = null;
 						$immediatelyInvokedCallable = TrinaryLogic::createMaybe();
 						$closureThisType = null;
+						$pureUnlessCallableIsImpureParameter = isset($pureUnlessCallableIsImpureParameters[$name]) && $pureUnlessCallableIsImpureParameters[$name];
```
`??`

staabm commented on src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php:
```diff
@@ -158,8 +174,8 @@ public function findFunctionReflection(string $functionName): ?NativeFunctionRef
 
-		if ($this->signatureMapProvider->hasFunctionMetadata($lowerCasedFunctionName)) {
-			$hasSideEffects = TrinaryLogic::createFromBoolean($this->signatureMapProvider->getFunctionMetadata($lowerCasedFunctionName)['hasSideEffects']);
+		if (isset($functionMetadata['hasSideEffects'])) {
```
`??`

staabm commented on stubs/arrayFunctions.stub:
```diff
@@ -9,6 +9,8 @@
  *
  * @return TReturn
+ *
+ * @pure-unless-callable-is-impure $two
```
did you check more stubs whether these need `@pure-unless-callable-is-impure`?

zonuexe committed:
Simplify pure-unless-callable-is-impure metadata reads

Address review nits: use null coalescing for the array<string, bool> parameter flags, a ternary for the hasSideEffects TrinaryLogic (keeping the absent -> Maybe branch), and split the negated method hasSideEffects read into named steps matching the ancestors loop below.

zonuexe committed:
Document the shape of the function metadata array

The value shape gained a second form (pureUnlessCallableIsImpureParameters), so describe both forms in the file-level comments of the source, the generator template and the generated file, add an inline @var before the array, and fix the generator's now-stale internal @var.

zonuexe committed:
Mark array_all() and array_any() as pure-unless-callable-is-impure

PHP 8.4's array_all() and array_any() are pure unless their callback is impure, like array_filter()/array_map().

zonuexe committed:
Honor pure-unless-callable-is-impure metadata for stub-reflected functions

Generic builtins like array_find()/array_find_key() are reflected via BetterReflection (PhpFunctionReflection), whose getCustomFunction() only read the @pure-unless-callable-is-impure PHPDoc tag and ignored functionMetadata.php - unlike the native-signature path. Merge the metadata's pureUnlessCallableIsImpureParameters in there too (PHPDoc tag wins). This makes array_find() actually honor its existing metadata entry and lets array_find_key() be added.

zonuexe committed:
Add lint version comment to the PHP 8.4 pure-unless fixture

RequiredPhpVersionCommentTest requires fixtures using version-specific features to declare a '// lint >= X.Y' comment so they are skipped on older PHP in CI.

zonuexe commented on src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php:
```diff
@@ -158,11 +174,9 @@ public function findFunctionReflection(string $functionName): ?NativeFunctionRef
-		}
+		$hasSideEffects = isset($functionMetadata['hasSideEffects'])
+			? TrinaryLogic::createFromBoolean($functionMetadata['hasSideEffects'])
+			: TrinaryLogic::createMaybe();
```
@staabm re your `??` suggestion at https://github.com/phpstan/phpstan-src/pull/3482#discussion_r3517629103

For the other `isset(...) && ...` spots I switched to `?? false` as you suggested. I kept an explicit ternary here, though:

```php
$hasSideEffects = isset($functionMetadata['hasSideEffects'])
    ? TrinaryLogic::createFromBoolean($functionMetadata['hasSideEffects'])
    : TrinaryLogic::createMaybe();
```

A plain `??` doesn't fit this one: the absent branch has to become `createMaybe()`, not `createFromBoolean(false)`. `createFromBoolean($functionMetadata['hasSideEffects'] ?? false)` would turn "metadata exists but has no `hasSideEffects` key" — e.g. `array_map()`, which only carries `pureUnlessCallableIsImpureParameters` — into `No` (definitely pure) instead of `Maybe`, silently dropping the impure-callback detection.

staabm commented on src/Reflection/Dummy/ChangedTypeMethodReflection.php:
```diff
@@ -168,6 +168,11 @@ public function isPure(): TrinaryLogic
 
+	public function getPureUnlessCallableIsImpureParameters(): array
+	{
+		return [];
```
```suggestion
		return $this->reflection->getPureUnlessCallableIsImpureParameters();
```

(seems we miss a test)

staabm commented on src/Reflection/Php/ClosureCallMethodReflection.php:
```diff
@@ -199,6 +199,11 @@ public function isPure(): TrinaryLogic
 
+	public function getPureUnlessCallableIsImpureParameters(): array
+	{
+		return [];
```
```suggestion
		return $this->nativeMethodReflection->getPureUnlessCallableIsImpureParameters();
```

staabm commented on src/Reflection/Php/PhpFunctionFromParserNodeReflection.php:
```diff
@@ -334,6 +339,11 @@ public function isPure(): TrinaryLogic
 
+	public function getPureUnlessCallableIsImpureParameters(): array
+	{
+		return [];
```
hmm does this miss the impl?

staabm commented on src/Reflection/Type/IntersectionTypeMethodReflection.php:
```diff
@@ -214,6 +214,11 @@ public function isPure(): TrinaryLogic
 
+	public function getPureUnlessCallableIsImpureParameters(): array
+	{
+		return [];
```
what about this?

staabm commented on src/Reflection/Type/UnionTypeMethodReflection.php:
```diff
@@ -171,6 +171,11 @@ public function isPure(): TrinaryLogic
 
+	public function getPureUnlessCallableIsImpureParameters(): array
+	{
+		return [];
```
and here?

zonuexe committed:
Honor @pure-unless-callable-is-impure on methods

Method support was entirely non-functional: PhpClassReflectionExtension only read the pure-unless map from built-in method metadata (never from a userland method's own PHPDoc tag), and PhpMethodReflection::getParameters() never passed the flag on to its PhpParameterReflection (the function path already did). Read the tag from the resolved PHPDoc and thread the flag through, and make the method-reflection wrappers (changed-type, closure-call, intersection, union, parser-node function) delegate getPureUnlessCallableIsImpureParameters() instead of returning [].

staabm commented on src/Reflection/Dummy/DummyConstructorReflection.php:
```diff
@@ -153,6 +153,11 @@ public function isPure(): TrinaryLogic
 		return TrinaryLogic::createYes();
 	}
 
+	public function getPureUnlessCallableIsImpureParameters(): array
```
also needs tests with constructor

zonuexe replied:
Good catch. You were right that method support was entirely non-functional, not only those stubs. Two gaps caused it.

- `PhpClassReflectionExtension` read the pure-unless map only from built-in method metadata, never from a userland method's own `@pure-unless-callable-is-impure` tag.
- `PhpMethodReflection::getParameters()` did not pass the flag on to `PhpParameterReflection`, while the function path already did.

I fixed both and made the wrappers you pointed at delegate. I also added a method regression test with pure, impure and opaque callbacks, including a union-typed receiver.

zonuexe committed:
Note the two paths that intentionally don't thread pure-unless yet

createNativeMethodVariant() (no built-in method carries the metadata) and Closure::call()'s parameters (a closure cannot carry the tag) leave the flag at its default; document why so a future maintainer knows it is deliberate, not an oversight.

zonuexe committed:
Honor @pure-unless-callable-is-impure on constructor calls

A 'new X(...)' where X's constructor is tagged only produced a fixed possibly-impure instantiation point, ignoring the callback's purity. NewHandler now applies the same pure-unless verdict used for function and method calls (extracted as a public static helper on SimpleImpurePoint): a pure callback drops the point, an impure one makes it certain. Constructors return void, so createFromVariant could not be reused directly (its void-return heuristic forces certain=true and skips the verdict).

zonuexe committed:
Drop the native mixed property type from the pure-unless fixture

RequiredPhpVersionCommentTest flags the non-php84 fixture because a native 'mixed' property type requires PHP 8.0 but the file carries no '// lint >=' comment. The property type is irrelevant to the test, so make it untyped to keep the fixture runnable on all supported PHP versions.

zonuexe committed:
Restructure NewHandler pure-unless branching to match createFromVariant

The previous compound condition ($verdict === null || !$verdict->yes()) produced mutants the constructor test did not kill in CI's Infection run. Use the same separate yes()/no() checks with an early return for the pure case that SimpleImpurePoint::createFromVariant already uses (and that Infection covers), so the branch is fully mutation-tested.

zonuexe committed:
Cover a union pure-callable|callable argument on a pure-unless constructor

staabm commented on src/Reflection/ExtendedParameterReflection.php:
```diff
@@ -36,4 +36,6 @@ public function getAllowedConstants(): ?ParameterAllowedConstants;
 	 */
 	public function checkAllowedConstants(array $constants): AllowedConstantsResult;
 
+	public function isPureUnlessCallableIsImpureParameter(): bool;
```
should return trinary.

quoting ondrey:

> I feel like on ExtendedParametersAcceptor, the method isPureUnlessCallableIsImpureParameter should return TrinaryLogic.
> Because you need to correctly set it on unions and intersection methods

staabm commented on tests/PHPStan/Rules/Pure/data/pure-unless-callable-is-impure.php:
```diff
@@ -0,0 +1,241 @@
+ * @param callable(int): int $f
+ * @param array<int> $arr
+ * @return array<int>
+ * @pure-unless-callable-is-impure $f
```
we need a test which shows that a function which is declared `@pure-unless-callable-is-impure $f` errors about impure side-effects when the impure statement/expression is outside the callable $f

staabm commented on tests/PHPStan/Rules/Pure/data/pure-unless-callable-is-impure.php:
```diff
@@ -0,0 +1,241 @@
+ * @param callable(int): int $f
+ * @param array<int> $arr
+ * @return array<int>
+ * @phpstan-pure-unless-callable-is-impure $f
```
If `@pure-unless-callable-is-impure`  is paired with `pure-callable` or `pure-Closure` , that should be a new kind of error (that the PHPdoc tag could be replaced with `@phpstan-pure` essentially

zonuexe committed:
Return TrinaryLogic from isPureUnlessCallableIsImpureParameter()

Per review: the parameter-level flag must be a TrinaryLogic so it composes
correctly on union and intersection method variants (combineAcceptors uses
->or(), mirroring isImmediatelyInvokedCallable). Origins convert the bool
from PHPDoc/metadata with TrinaryLogic::createFromBoolean(); the consumption
in SimpleImpurePoint skips only parameters whose flag is ->no().

zonuexe committed:
Check the body of @pure-unless-callable-is-impure functions for purity

A function/method declared @pure-unless-callable-is-impure $f is pure except
for the flagged callable, so its body is now checked by FunctionPurityCheck:
impure statements outside the callable (e.g. echo) are reported, while the
flagged callable's own invocation is exempt. Thread the parameter flags through
enterFunction() so the check sees them on plain functions too.

zonuexe committed:
Report redundant @pure-unless-callable-is-impure on a pure callable

Per review: when @pure-unless-callable-is-impure $f is paired with a pure
callable type (pure-callable / pure-Closure) for $f, the tag is redundant
because the callback is always pure, so the function/method can be marked
@phpstan-pure instead. FunctionPurityCheck now reports this.

ondrejmirtes commented on src/Reflection/ExtendedMethodReflection.php:
```diff
@@ -69,6 +69,11 @@ public function isBuiltin(): TrinaryLogic|bool;
 	public function isPure(): TrinaryLogic;
 
+	/**
+	 * @return array<string, bool>
```
Should still be TrinaryLogic here. When you have UnionMethodReflection or IntersectionMethodReflection (and the methods in question have different PHPDoc tags involved, you need TrinaryLogic to properly model the behaviour). Also add some tests.

ondrejmirtes reviewed (changes requested):
One more thing - in MethodSignatureRule, we need to enforce covariance for these methods. Check around `$this->reportMethodPurityOverride`.

If a method has `@phpstan-pure` or `@phpstan-pure-unless-callable-is-impure`, the child method cannot be `@phpstan-impure`.

Also are there same tests around `@phpstan-pure-unless-callable-is-impure` being inherited implicitly? (The child method does not need to mention the parameter but if parent method has it, it should also apply to child. Even when the parameter is renamed.)

zonuexe committed:
Return TrinaryLogic from getPureUnlessCallableIsImpureParameters()

Per review (ondrejmirtes): the method/function-level map must return
array<string, TrinaryLogic> so UnionTypeMethodReflection and
IntersectionTypeMethodReflection can model Maybe when constituent methods
disagree on the tag (new MergedPureUnlessCallableIsImpureParameters helper).
combineAcceptors() now merges the parameter-level flag with equals-or-Maybe
instead of ->or(), so a union receiver yields a Maybe flag. FunctionPurityCheck
drives its flagged-parameter set off the map getter. Adds union and inherited
(incl. renamed-parameter) regression tests.

zonuexe committed:
Report impure methods overriding @pure-unless-callable-is-impure methods

Per review (ondrejmirtes): MethodSignatureRule now enforces purity covariance
for the tag - a child method marked @phpstan-impure cannot override a parent
method carrying @pure-unless-callable-is-impure (identifier
method.impureOverridePureUnlessCallable), mirroring the existing pure-override
check. Gated by reportMethodPurityOverride.

zonuexe committed:
Add a direct unit test for resolvePureUnlessCallableIsImpureVerdict()

Kills the TrinaryLogicMutator mutant on the parameter-flag check
(->no() -> !->yes()) deterministically, without going through the
analyser: a Maybe flag must be processed like Yes, so with the callback
omitted the verdict is Yes, while the mutant skips the parameter and
yields null.

zonuexe committed:
Make SimpleImpurePointTest a self-contained unit test

Extend plain TestCase and pass a mocked Scope instead of booting the DI
container and creating a real analyser Scope. The verdict method never
consults the scope when no arguments are passed, so this keeps the test
fully isolated (avoids interfering with other tests' shared state) while
still killing the parameter-flag mutant.

zonuexe committed:
Share the pure-unless-callable verdict-to-certainty decision

Extract the verdict->yes()/->no() branching from createFromVariant() and
NewHandler into SimpleImpurePoint::applyPureUnlessCallableIsImpureVerdict(),
covered by a direct unit test. This removes the duplicated logic and lets the
unit test deterministically kill the mutants on those branches instead of
relying on coverage attribution from the rule tests.

zonuexe committed:
Revert the direct mutation-testing unit test and verdict helper

Reverts 33f825ab7, f0cf3d496 and 4859c2f99. Local Infection with the exact
custom TrinaryLogicMutator kills every mutant on the changed lines (MSI 100%),
so the remaining CI Mutation Testing failure is a coverage-attribution issue in
the parallel run rather than a missing test - the added unit test and the
verdict-to-certainty helper did not change that, so they are dropped to keep the
PR focused on the reviewed changes.

staabm commented:
As a followup it would be great to look into phpstan 1st party extension packages and also often used packages in the php ecosystem, which could utilize the new phpdoc
